File: AOC2020/d17/a.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

const (
	sx, sy, sz = 11, 11, 11
	lim        = 30
)

type Coor struct {
	x, y, z int
}

func getFriends(x, y, z int) []Coor {
	inRange := func(z int) bool { return z >= 0 && z < lim }
	arr := make([]Coor, 0)
	for i := -1; i < 2; i++ {
		for j := -1; j < 2; j++ {
			for k := -1; k < 2; k++ {
				if i == 0 && j == 0 && k == 0 { continue }
				nx, ny, nz := x+i, y+j, z+k
				if inRange(nx) && inRange(ny) && inRange(nz) {
					arr = append(arr, Coor{nx, ny, nz})
				}
			}
		}
	}
	return arr
}

func runCycle(grid [][][]bool) int {
	countActive := func(coor []Coor) int {
		cnt := 0
		for _, c := range coor {
			if grid[c.x][c.y][c.z] { cnt++ }
		}
		return cnt
	}
	flipped := make([]Coor, 0)
	tot := 0
	for i := 0; i < lim; i++ {
		for j := 0; j < lim; j++ {
			for k := 0; k < lim; k++ {
				friends := getFriends(i, j, k)
				active := countActive(friends)
				if grid[i][j][k] {
					tot++
					if active == 2 || active == 3 {
					} else {
						flipped = append(flipped, Coor{i, j, k})
					}
				}
				if !grid[i][j][k] && active == 3 {
					flipped = append(flipped, Coor{i, j, k})
				}
			}
		}
	}
	for _, c := range flipped {
		grid[c.x][c.y][c.z] = !grid[c.x][c.y][c.z]
		if grid[c.x][c.y][c.z] {
			tot++
		} else {
			tot--
		}
	}
	return tot
}

func main() {
	bytes, err := ioutil.ReadFile("input")
	if err != nil {
		fmt.Println(err)
		return
	}
	mat := strings.Split(string(bytes), "\n")
	fmt.Println(mat)

	grid := make([][][]bool, lim)

	for i := 0; i < lim; i++ {
		grid[i] = make([][]bool, lim)
		for j := 0; j < lim; j++ {
			grid[i][j] = make([]bool, lim)
		}
	}

	for i := 0; i < len(mat); i++ {
		for j := 0; j < len(mat[0]); j++ {
			if mat[i][j] == '#' { grid[i+sx][j+sy][sz] = true }
		}
	}
	for times := 0; times < 6; times++ {
		active := runCycle(grid)
		fmt.Println("cyc", times, "act", active)
	}

}

File: AOC2020/d17/a.cpp
#include <iostream>
#include <vector>

using namespace std;
 
#define fr(i, st, n) for(int i = (int )st; i < (int )n; i++)
#define rfr(i, en, st) for(int i = (int )en; i >= (int )st; i--)
#define pb push_back
#define sz(a) (int)(a).size()
 
typedef vector< int> vi;

 // START FROM HERE

const int lim=30;
const int sx=11,sy=11,sz=11;
bool grid[lim][lim][lim];

struct Coo
{
 int x,y,z;
 // Coo(){}
 Coo(int x,int y,int z):x(x),y(y),z(z){}
};

int cycle(){
 std::vector<Coo> v;
 auto f=[](int x){
	return x>=0&&x<lim;
 };
 int tot=0;
 fr(x,0,lim){
	fr(y,0,lim){
	 fr(z,0,lim){
		int cnt=0;
		fr(dx,-1,2)fr(dy,-1,2)fr(dz,-1,2){
		 if(dx==0&&dy==0&&dz==0)continue;
		 if(f(dx+x)&&f(dy+y)&&f(dz+z)&&grid[x+dx][y+dy][z+dz]){
			++cnt;
		 }
		}
		if(grid[x][y][z]){
		 tot++;
		 if (cnt==2||cnt==3){

		 }else{
			v.pb(Coo(x,y,z));
		 }
		}
		if(!grid[x][y][z]&&cnt==3){
		 v.pb(Coo(x,y,z));
		}

	 }
	}
 }
 for(auto c:v){
	grid[c.x][c.y][c.z] = !grid[c.x][c.y][c.z];
	if (grid[c.x][c.y][c.z]) tot++;
	else tot--;
 }
 return tot;


}

int main(int argc, char const *argv[])
{
 fr(i,0,lim)fr(j,0,lim)fr(k,0,lim){
	grid[i][j][k]=false;
 }
 // std::vector<string> mat({".#.","..#","###"});
 std::vector<string> mat({"..#....#",
"##.#..##",
".###....",
"#....#.#",
"#.######",
"##.#....",
"#.......",
".#......"});
 fr(i,0,sz(mat)){
	fr(j,0,sz(mat[0])){
	 if (mat[i][j]=='#'){
	 grid[i+sx][j+sy][sz]=true;
	 }
	}
 }
 fr(i,0,6)
 cout<<cycle()<<endl;


 return 0;
}





// {"..#....#",
// "##.#..##",
// ".###....",
// "#....#.#",
// "#.######",
// "##.#....",
// "#.......",
// ".#......"}